lib/broccoli/ember-app.js

```diff
@@ -1752,12 +1738,26 @@ class EmberApp {
     @return {Tree}                  Merged tree for this application
    */
   toTree(additionalTrees) {
-    let tree = mergeTrees(this.toArray().concat(additionalTrees || []), {
+    let trees = this.toArray().concat(additionalTrees || []);
+    let isPackageHookSupplied = typeof this.package === 'function' || typeof this.options.package === 'function';
+    let packageFn = this.options.package || this.package;
+
+    let fullTree = mergeTrees(trees, {
       overwrite: true,
-      annotation: 'TreeMerger (allTrees)',
+      annotation: 'Full Application',
     });
 
-    return this.addonPostprocessTree('all', tree);
+    fullTree = this._debugTree(fullTree, 'prepackage');
+
+    if (experiments.PACKAGER) {
+      if (isPackageHookSupplied) {
+        return this.addonPostprocessTree('all', packageFn.call(this, fullTree));
```

Comment:
Maybe we should do `packagedTree = packageFn.call(this, fullTree)` here, then we can do `return this.addonPostprocessTree('all', packagedTree || this.defaultPackage(fullTree))` below.